File: LeetCodeCPlusPlus/py20.cpp
#include <map>
#include <stack>
#include <string>
using namespace std;
// 括号匹配，栈问题
class Solution {
public:
    bool isValid(string s) {
        map<char,char> lookup = {
                {'[', ']'},
                {'{', '}'},
                {'(', ')'}
        };
        stack<char> mystack;
        for(auto i:s) {
            if (lookup.count(i)==1){
                mystack.push(i);
                continue;
            } 
            else if(!mystack.empty() && lookup[mystack.top()] == i) 
                mystack.pop();
            else
                return false;
        }
    return mystack.empty();
   }
};

File: LeetCodePython/py20.py

class Solution:
    # 有效符号，栈匹配
    def isValid(self, s: str) -> bool:
        ans = ['#',]
        s = s + '#'
        for i in range(len(s)):
            if ans[-1] == '#' and s[i] == '#' or \
               ans[-1] == '(' and s[i] == ')' or \
               ans[-1] == '{' and s[i] == '}' or \
               ans[-1] == '[' and s[i] == ']':
                ans.pop()
            else:
                ans.append(s[i])
        return not ans

class Solution2:
    def isValid(self, s: str) -> bool:
        lookup = {'{': '}',  '[': ']', '(': ')'}
        stack = []
        for c in s:
            if c in lookup:
                stack.append(c)
                continue
            if stack and lookup[stack[-1]]==c:
                stack.pop()
            else:
                return False
        return not stack
    


                    
        